File: src/downloads_organizer/locking.py
from __future__ import annotations

import fcntl
from contextlib import contextmanager
from pathlib import Path
from typing import Iterator


@contextmanager
def app_lock(data_dir: Path) -> Iterator[None]:
    data_dir.mkdir(parents=True, exist_ok=True)
    lock_path = data_dir / "organizer.lock"
    with lock_path.open("a+") as handle:
        try:
            fcntl.flock(handle.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError as exc:
            raise RuntimeError("另一个 Downloads Organizer 进程正在运行") from exc
        try:
            yield
        finally:
            fcntl.flock(handle.fileno(), fcntl.LOCK_UN)

File: Sources/TopicTidyCore/Support/Locking.swift
import Foundation

/// Single-writer guard shared by the CLI, the GUI and the daily task.
public enum AppLock {
    public static func withLock<T>(_ dataDir: URL, _ body: () throws -> T) throws -> T {
        try FileManager.default.createDirectory(atPath: dataDir.path, withIntermediateDirectories: true)
        let lockPath = PyPath.join(dataDir, "organizer.lock").path
        let descriptor = open(lockPath, O_RDWR | O_CREAT, 0o644)
        guard descriptor >= 0 else {
            throw OrganizerError("无法打开应用锁：\(lockPath)")
        }
        defer { close(descriptor) }
        if flock(descriptor, LOCK_EX | LOCK_NB) != 0 {
            throw OrganizerError("另一个 Downloads Organizer 进程正在运行")
        }
        defer { flock(descriptor, LOCK_UN) }
        return try body()
    }
}
